### PR TITLE
test(determinism): add determinism tests

### PR DESCRIPTION
**Pull request by ZhiyuLi-Nvidia**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?

Adds determinism test infrastructure for Megatron-LM:

1. **`megatron/training/determinism.py`** — extracts `set_determinism_env_vars` and `apply_determinism_to_args` for `validate_args` + reuse from tests.

2. **Bit-exact correctness suite** (`tests/unit_tests/determinism/`) —
   `BitExactRunner` parametrizes (GPTModel × HybridModel × TransformerBlock) × (TP / EP / FSDP / PP / VPP composites + FP8/FP4 recipes). Two runs of the same cell must produce bit-identical outputs and gradients. Full matrix ~3 min pytest / ~5 min slurm on 8×H100.

3. **Determinism perf check**
   (`tests/performance_tests/shell_test_utils/determinism/`) — wraps `pretrain_gpt.py` twice (det + nondet) under nsys, prints a 3-table per-NVTX leaderboard (forward / backward / op-level), and fails CI if det/nondet step-time ratio > 1.25×. Routed as `gpt/determinism_perf` in the integration matrix.

```
=== forward — mcore module ranges (top 20 by |det - nondet|) ===
Range                                                                               det ms   nondet ms    delta ms    delta %
-----------------------------------------------------------------------------------------------------------------------------
:megatron.core.transformer.transformer_layer._forward_attention.self_attention     409.957     390.898     +19.060      +4.88
:megatron.core.transformer.transformer_layer._forward_mlp.mlp                      199.910     181.300     +18.610     +10.26
:megatron.core.transformer.mlp.forward.linear_fc1                                   55.073      44.514     +10.559     +23.72
:megatron.core.transformer.transformer_layer._forward_post_mlp.mlp_bda              52.849      42.983      +9.867     +22.95
:megatron.core.transformer.transformer_layer._forward_attention.self_attn_bda       60.718      50.881      +9.836     +19.33
:megatron.core.transformer.attention.forward.qkv                                    74.025      64.415      +9.611     +14.92
:megatron.core.transformer.attention.forward.core_attention                        115.101     106.197      +8.903      +8.38
....

=== backward — autograd engine ranges (top 20 by |det - nondet|) ===
Range                                                                                 det ms   nondet ms    delta ms    delta %
-------------------------------------------------------------------------------------------------------------------------------
:_LayerNormLinearBackward                                                            154.887     130.923     +23.964     +18.30
:autograd::engine::evaluate_function: CompiledFunctionBackward                        85.340      61.626     +23.714     +38.48
:autograd::engine::evaluate_function: _LayerNormLinearBackward                       170.793     147.138     +23.655     +16.08
:CompiledFunctionBackward                                                             80.341      56.708     +23.634     +41.68
:autograd::engine::evaluate_function: _VocabParallelCrossEntropyBackward              31.404      13.724     +17.680    +128.83
:_VocabParallelCrossEntropyBackward                                                   29.495      11.931     +17.563    +147.20
:autograd::engine::evaluate_function: IndexBackward0                                  11.426           -     +11.426          -
:autograd::engine::evaluate_function: _LinearBackward                                 71.758      61.127     +10.631     +17.39
:_LinearBackward                                                                      65.801      55.296     +10.505     +19.00
:IndexBackward0                                                                        9.677           -      +9.677          -
....

=== op-level — aten / NCCL / kernels (top 20 by |det - nondet|) ===
Range                                                                                 det ms   nondet ms    delta ms    delta %
-------------------------------------------------------------------------------------------------------------------------------
:aten::fill_                                                                         144.613       4.570    +140.042   +3064.27
:aten::empty                                                                         131.060      35.005     +96.055    +274.40
:aten::empty_strided                                                                  86.055       8.258     +77.797    +942.10
:aten::_index_put_impl_                                                               30.761       5.995     +24.767    +413.14
:aten::item                                                                           78.065      56.393     +21.672     +38.43
:aten::_local_scalar_dense                                                            76.898      55.248     +21.650     +39.19
:aten::is_nonzero                                                                     59.699      38.317     +21.383     +55.80
:aten::empty_like                                                                     29.083       8.609     +20.474    +237.81
:_LayerNormLinear                                                                     80.321      60.092     +20.228     +33.66
...
:## Call CompiledFxGraph fomircajdh247tezeg7zugaol2mm5xpna3vlaz7sa4c3oemupsii ##       8.501           -      +8.501          -
:## Call CompiledFxGraph fkfi3f66hdefg2ffuf54lx5uhqekl7kgmb72hqmrmadyw6d6ebui ##       8.131           -      +8.131          -
:## Call CompiledFxGraph f2mtwsp4qmnpib3fi4nzll55ea7inickj46b6tjk44mzarn6pano ##       8.054           -      +8.054          -

step_time iter=7: det=107.30ms nondet=94.50ms ratio=1.14x (threshold 1.25x)
```

4. **Stress utilities** — `RacingStreams` and `CudaSleepJitter` surface
   latent collective-ordering bugs as bit-exact failures.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by maanug-nv — APPROVED** [review]

**Review by ko3n1g — APPROVED** [review]

File: megatron/training/determinism.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    # 4. Override tp_comm_overlap. ``warn_rank_0`` (not print_rank_0) so the
+    #    override is capturable via ``pytest.warns`` / ``-W error``.
+    if args.tp_comm_overlap:
+        # Lazy import — warn_rank_0 lives in training.utils which has heavier
+        # dependencies than this module.
+        from megatron.training.utils import warn_rank_0
+
+        warn_rank_0("Disabling tp_comm_overlap for deterministic mode.")
+        args.tp_comm_overlap = False
```
Comment (jaredcasper) [inline megatron/training/determinism.py]: Can they not do this themselves? I would much prefer to error out and tell them they need to turn it off, than to silently turn it off here. (A warning in the log is "silent").

File: megatron/training/determinism.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    These env vars are captured by their respective libraries at first use
+    (NCCL at communicator init, cuBLAS at handle creation, TE at first
+    attention forward), so the call must happen BEFORE any of those events.
+    Uses ``setdefault`` so any value the launcher has already exported
+    wins — defense in depth: in the test process pytest may import another
+    module that triggers CUDA-context creation before this package loads,
+    in which case the Python-side setdefault is too late and the launcher's
+    shell-side export is what actually takes effect.
+    """
+    os.environ.setdefault("NCCL_ALGO", "Ring")
```
Comment (jaredcasper) [inline megatron/training/determinism.py]: So if they have alrady set NCCL_ALGO to something other than "Ring" that wins, and the run won't be deterministic? Should we error out if they have set NCCL_ALGO to something other than "Ring"?

Comment (jaredcasper) [reply megatron/training/determinism.py]: Oh nevermind, I see below that is part of the test.

File: megatron/training/determinism.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    # 2. NCCL_ALGO sanity check on a launcher-supplied value. Tree is
+    #    intentionally excluded: its intra-node chain reduction order is not
+    #    user-controllable and multi-node inter-tree topology can vary
+    #    across runs without a pinned topology file, so we cannot vouch
+    #    for it as deterministic. ``^NVLS`` is kept because banning NVLS
+    #    on hardware that exposes it is a legitimate user choice; the user
+    #    is responsible for picking a fallback that's deterministic on
+    #    their stack. Comma-separated lists are valid NCCL syntax.
+    accepted_tokens = {"Ring", "CollnetDirect", "CollnetChain", "^NVLS"}
+    nccl_algo = os.environ.get("NCCL_ALGO")
```
Comment (jaredcasper) [inline megatron/training/determinism.py]: Should we also check the other two set in set_determinism_env_vars?

File: megatron/training/determinism.py (hunk 1 above)
Comment (ZhiyuLi-Nvidia) [reply megatron/training/determinism.py]: Thank you. 

Agreed. Switched from silent config override to an assertion that errors out.

File: megatron/training/determinism.py (hunk 3 above)
Comment (ZhiyuLi-Nvidia) [reply megatron/training/determinism.py]: Yeap,  both `CollnetDirect` and `CollnetChain` have been validated deterministic. Update comments for better understanding.

Comment (jaredcasper) [reply megatron/training/determinism.py]: Sorry, I mean the other env vars. Shouldn't we also check here that NVTE_ALLOW_NONDETERMINISTIC_ALGO is "0" and CUBLAS_WORKSPACE_CONFIG is ":4096:8" (or something that works).

Comment (ZhiyuLi-Nvidia) [reply megatron/training/determinism.py]: Thank you @jaredcasper. Gotcha, sanity check updated.

apply_determinism_to_args now applies the same "unset OK, set-and-invalid errors" pattern to all four env vars:

- NCCL_ALGO — comma-separated subset of {Ring, CollnetDirect, CollnetChain, ^NVLS}
- [added] NVTE_ALLOW_NONDETERMINISTIC_ALGO — "0" only
- [added] CUBLAS_WORKSPACE_CONFIG — `:4096:8` or `:16:8` (both are deterministic per NVIDIA cuBLAS reproducibility docs; differ only in workspace size)
- [added] MAMBA_DETERMINISTIC — '1'. Mamba SSM module auto-follows torch.are_deterministic_algorithms_enabled() when unset, so we only reject an explicit non-deterministic override.

**Review by jaredcasper — APPROVED** [review]
